source/particle/integrator/rk_4.cc

```diff
@@ -0,0 +1,186 @@
+      {
+        // TODO: If we finished integration we do not need to transfer integrator data. Return 0 in that case.
+
+        return 4*dim*sizeof(double);
+      }
+
+      template <int dim>
+      void
+      RK4<dim>::read_data(const void *&data,
+                          const types::particle_index id_num)
+      {
+        // TODO: If we finished integration we do not need to transfer integrator data. Return early in that case.
+
+        const double *integrator_data = static_cast<const double *> (data);
+
+        // Read location data
+        for (unsigned int i=0; i<dim; ++i)
+          loc0[id_num](i) = *integrator_data++;
+
+        // Read k1, k2 and k3
+        for (unsigned int i=0; i<dim; ++i)
+          k1[id_num][i] = *integrator_data++;
+
+        for (unsigned int i=0; i<dim; ++i)
+          k2[id_num][i] = *integrator_data++;
+
+        for (unsigned int i=0; i<dim; ++i)
+          k3[id_num][i] = *integrator_data++;
+
+        data = static_cast<const void *> (integrator_data);
+      }
+
+      template <int dim>
+      void
+      RK4<dim>::write_data(void *&data,
+                           const types::particle_index id_num) const
+      {
+        // TODO: If we finished integration we do not need to transfer integrator data. Return early in that case.
+
+        double *integrator_data = static_cast<double *> (data);
+
+        // Write location data
+        typename std::map<types::particle_index, Point<dim> >::const_iterator it = loc0.find(id_num);
+        for (unsigned int i=0; i<dim; ++i,++integrator_data)
+          *integrator_data = it->second(i);
+
+        // Write k1, k2 and k3
+        typename std::map<types::particle_index, Tensor<1,dim> >::const_iterator it_k = k1.find(id_num);
+        for (unsigned int i=0; i<dim; ++i,++integrator_data)
+          *integrator_data = it_k->second[i];
+
+        it_k = k2.find(id_num);
+        for (unsigned int i=0; i<dim; ++i,++integrator_data)
+          *integrator_data = it_k->second[i];
+
+        it_k = k3.find(id_num);
+        for (unsigned int i=0; i<dim; ++i,++integrator_data)
+          *integrator_data = it_k->second[i];
+
+        data = static_cast<void *> (integrator_data);
+      }
+    }
+  }
+}
+
+
+// explicit instantiations
+namespace aspect
+{
+  namespace Particle
+  {
+    namespace Integrator
+    {
+      ASPECT_REGISTER_PARTICLE_INTEGRATOR(RK4,
+                                          "rk4",
+                                          "Runge Kutta fourth order integrator, where "
+                                          "y_{n+1} = y_n + (1/6)*k1 + (1/3)*k2 + (1/3)*k3 + (1/6)*k4 "
+                                          "and k1, k2, k3, k4 are defined as usual. "
+                                          "This scheme requires storing the original location "
+                                          "and intermediate k1, k2, k3 values, so the "
+                                          "read/write_data functions reflect this.")
```

Comment:
This sentence is not useful to users. It only describes an implementation detail.